### ISSUE TITLE
XF Null exception at  Prism.Common.PageUtilities.DestroyPage 

### ISSUE THREAD

**Issue by snowppy** [PrismLibrary/Prism#966]
I have started getting the following exception whenever the hardware button is used to navigate back from a page after i updated my xamarin forms version. Right now I am on  4.3.1.1  but it started popping up in the previous version as well pretty much for the last two weeks. The exception doesn't kill the app and I only get it in the App hockey analytic s, so I am not sure it should have any priority though I get it on a lot of pages causing the  issue.

`Xamarin caused by: android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.AggregateException: One or more errors occurred. ---> System.Exception: Cannot destroy LwdApp.Pages.Location.CreateLocationPage: Object reference not set to an instance of an object
  at Prism.Common.PageUtilities.DestroyPage (Xamarin.Forms.Page page) [0x00063] in <87556e649730407189f11a7a6bcc0b81>:0 
  at Prism.Behaviors.NavigationPageSystemGoBackBehavior.NavigationPage_Popped (System.Object sender, Xamarin.Forms.NavigationEventArgs e) [0x00031] in <87556e649730407189f11a7a6bcc0b81>:0 
  at Xamarin.Forms.NavigationPage+<Xamarin-Forms-INavigationPageController-PopAsyncInner>d__63.MoveNext () [0x0016b] in <59b03de64f0e485e8c9d8ead0b747e17>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at Xamarin.Forms.NavigationPage+<PopAsync>d__38.MoveNext () [0x001b9] in <59b03de64f0e485e8c9d8ead0b747e17>:0 
   --- End of inner exception stack trace ---
  at Xamarin.Forms.NavigationPage+<>c.<SafePop>b__93_0 (System.Threading.Tasks.Task`1[TResult] t) [0x0000e] in <59b03de64f0e485e8c9d8ead0b747e17>:0 
  at System.Threading.Tasks.ContinuationTaskFromResultTask`1[TAntecedentResult].InnerInvoke () [0x00027] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Threading.Tasks.Task.Execute () [0x00016] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.AggregateException: One or more errors occurred. ---> System.Exception: Cannot destroy LwdApp.Pages.Location.CreateLocationPage: Object reference not set to an instance of an object
  at Prism.Common.PageUtilities.DestroyPage (Xamarin.Forms.Page page) [0x00063] in <87556e649730407189f11a7a6bcc0b81>:0 
  at Prism.Behaviors.NavigationPageSystemGoBackBehavior.NavigationPage_Popped (System.Object sender, Xamarin.Forms.NavigationEventArgs e) [0x00031] in <87556e649730407189f11a7a6bcc0b81>:0 
  at Xamarin.Forms.NavigationPage+<Xamarin-Forms-INavigationPageController-PopAsyncInner>d__63.MoveNext () [0x0016b] in <59b03de64f0e485e8c9d8ead0b747e17>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at Xamarin.Forms.NavigationPage+<PopAsync>d__38.MoveNext () [0x001b9] in <59b03de64f0e485e8c9d8ead0b747e17>:0 
   --- End of inner exception stack trace ---
  at Xamarin.Forms.NavigationPage+<>c.<SafePop>b__93_0 (System.Threading.Tasks.Task`1[TResult] t) [0x0000e] in <59b03de64f0e485e8c9d8ead0b747e17>:0 
  at System.Threading.Tasks.ContinuationTaskFromResultTask`1[TAntecedentResult].InnerInvoke () [0x00027] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Threading.Tasks.Task.Execute () [0x00016] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
---> (Inner Exception #0) System.Exception: Cannot destroy LwdApp.Pages.Location.CreateLocationPage: Object reference not set to an instance of an object
  at Prism.Common.PageUtilities.DestroyPage (Xamarin.Forms.Page page) [0x00063] in <87556e649730407189f11a7a6bcc0b81>:0 
  at Prism.Behaviors.NavigationPageSystemGoBackBehavior.NavigationPage_Popped (System.Object sender, Xamarin.Forms.NavigationEventArgs e) [0x00031] in <87556e649730407189f11a7a6bcc0b81>:0 
  at Xamarin.Forms.NavigationPage+<Xamarin-Forms-INavigationPageController-PopAsyncInner>d__63.MoveNext () [0x0016b] in <59b03de64f0e485e8c9d8ead0b747e17>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <da4450a5a5a948c3b1015f0ba717bfc0>:0 
  at Xamarin.Forms.NavigationPage+<PopAsync>d__38.MoveNext () [0x001b9] in <59b03de64f0e485e8c9d8ead0b747e17>:0 <---
<---`

**Comment by brianlagunas** [comment]
This is most likely a result of you trying to use a null object in your Destroy() method.  Destroy cannot be called on a null object, so if a null exception is being thrown, it most likely exists in your code.  Please check for nulls before trying to call methods on your objects in Destroy.

If you can create a repro that proves it is happening in Prism, I will reopen and address the issue.  Let me know what you find.

**Comment by tonholis** [comment]
Hi @brianlagunas !
I'm using the latest version of Xamarin Prism (on April 14), and I get the same error. But in my case, it occurred in a custom Entry Behavior that I created and I didn't check for null on TextChangedEventArgs.NewTextValue. My fault.

But I noticed that on PageDestroy the BindingContext is changed and bubbled to all controls in my page. Is there any way to avoid this? Check the stack trace:

```
  at MyApp.Controls.Behaviors.MaskBehavior.HandleTextChanged (System.Object sender, Xamarin.Forms.TextChangedEventArgs e)
  at (wrapper delegate-invoke) System.EventHandler`1[Xamarin.Forms.TextChangedEventArgs]:invoke_void_object_TEventArgs (object,Xamarin.Forms.TextChangedEventArgs)
  at Xamarin.Forms.Entry.OnTextChanged (Xamarin.Forms.BindableObject bindable, System.Object oldValue, System.Object newValue)
  at Xamarin.Forms.BindableObject.SetValueActual (Xamarin.Forms.BindableProperty property, Xamarin.Forms.BindableObject+BindablePropertyContext context, System.Object value, System.Boolean currentlyApplying, Xamarin.Forms.BindableObject+SetValueFlags attributes, System.Boolean silent)
  at Xamarin.Forms.BindableObject.SetValueCore (Xamarin.Forms.BindableProperty property, System.Object value, Xamarin.Forms.BindableObject+SetValueFlags attributes, Xamarin.Forms.BindableObject+SetValuePrivateFlags privateAttributes)
  at Xamarin.Forms.BindingExpression.ApplyCore (System.Object sourceObject, Xamarin.Forms.BindableObject target, Xamarin.Forms.BindableProperty property, System.Boolean fromTarget)
  at Xamarin.Forms.BindingExpression.Apply (System.Object sourceObject, Xamarin.Forms.BindableObject target, Xamarin.Forms.BindableProperty property)
  at Xamarin.Forms.Binding.Apply (System.Object newContext, Xamarin.Forms.BindableObject bindObj, Xamarin.Forms.BindableProperty targetProperty)
  at Xamarin.Forms.BindableObject.ApplyBindings (System.Boolean skipBindingContext)
  at Xamarin.Forms.BindableObject.ApplyBindings ()
  at Xamarin.Forms.BindableObject.SetInheritedBindingContext (Xamarin.Forms.BindableObject bindable, System.Object value)
  at Xamarin.Forms.Element.SetChildInheritedBindingContext (Xamarin.Forms.Element child, System.Object context)
  at Xamarin.Forms.Element.OnBindingContextChanged ()
  at Xamarin.Forms.View.OnBindingContextChanged ()
  at Xamarin.Forms.BindableObject.SetInheritedBindingContext (Xamarin.Forms.BindableObject bindable, System.Object value)
  at Xamarin.Forms.Element.SetChildInheritedBindingContext (Xamarin.Forms.Element child, System.Object context)
  at Xamarin.Forms.Element.OnBindingContextChanged ()
  at Xamarin.Forms.View.OnBindingContextChanged ()
  at Xamarin.Forms.Grid.OnBindingContextChanged ()
  at Xamarin.Forms.BindableObject.SetInheritedBindingContext (Xamarin.Forms.BindableObject bindable, System.Object value)
  at Xamarin.Forms.Element.SetChildInheritedBindingContext (Xamarin.Forms.Element child, System.Object context)
  at Xamarin.Forms.Element.OnBindingContextChanged ()
  at Xamarin.Forms.View.OnBindingContextChanged ()
  at Xamarin.Forms.BindableObject.SetInheritedBindingContext (Xamarin.Forms.BindableObject bindable, System.Object value)
  at Xamarin.Forms.Element.SetChildInheritedBindingContext (Xamarin.Forms.Element child, System.Object context)
  at Xamarin.Forms.Element.OnBindingContextChanged ()
  at Xamarin.Forms.View.OnBindingContextChanged ()
  at Xamarin.Forms.BindableObject.SetInheritedBindingContext (Xamarin.Forms.BindableObject bindable, System.Object value)
  at Xamarin.Forms.Element.SetChildInheritedBindingContext (Xamarin.Forms.Element child, System.Object context)
  at Xamarin.Forms.TemplatedPage.SetChildInheritedBindingContext (Xamarin.Forms.Element child, System.Object context)
  at Xamarin.Forms.Element.OnBindingContextChanged ()
  at Xamarin.Forms.Page.OnBindingContextChanged ()
  at Xamarin.Forms.ContentPage.OnBindingContextChanged ()
  at Xamarin.Forms.BindableObject.BindingContextPropertyChanged (Xamarin.Forms.BindableObject bindable, System.Object oldvalue, System.Object newvalue)
  at Xamarin.Forms.BindableObject.SetValueActual (Xamarin.Forms.BindableProperty property, Xamarin.Forms.BindableObject+BindablePropertyContext context, System.Object value, System.Boolean currentlyApplying, Xamarin.Forms.BindableObject+SetValueFlags attributes, System.Boolean silent)
  at Xamarin.Forms.BindableObject.SetValueCore (Xamarin.Forms.BindableProperty property, System.Object value, Xamarin.Forms.BindableObject+SetValueFlags attributes, Xamarin.Forms.BindableObject+SetValuePrivateFlags privateAttributes)
  at Xamarin.Forms.BindableObject.SetValue (Xamarin.Forms.BindableProperty property, System.Object value, System.Boolean fromStyle, System.Boolean checkAccess)
  at Xamarin.Forms.BindableObject.SetValue (Xamarin.Forms.BindableProperty property, System.Object value)
  at Xamarin.Forms.BindableObject.set_BindingContext (System.Object value)
  at Prism.Common.PageUtilities.DestroyPage (Xamarin.Forms.Page page)
```

**Comment by brianlagunas** [comment]
@tonholis When a Page is removed from the navigation stack (destroyed) the BindingContext is set to null to avoid any memory leaks that XF is notorious for having.  Currently there is no way to change this behavior.  If we didn't null it out, you would have a memory leak because your behavior has an event handler for an event of a control in your Page.  So actually, this worked exactly as it was supposed to.  Prism let you know that you had a potential memory leak with that event handler :)

**Comment by brianlagunas** [comment]
By the way, I would suggest adding IDestructible to your View and removing all behaviors you have defined in your Page.

**Comment by tonholis** [comment]
Great! Makes sense now.
Thanks for the explanation and your suggestion.  I'll review it.

**Comment by odahan** [comment]
I 'm just having same problem when navigating. All is ok until I play with fields in View that are bound to the viewmodel (two pickers, "new xamarin model" having binding). VS, Windows, Xamarin and Prism are up to date at the date of this message.
Until I'm not touching the pickers, all is ok I can navigate. Once a picker has been used to select a value the navigation is returning the error (cannot destroy...).
I'm using a master/detail page for navigation and this is working (almost) well (navigating to any new location resets the history and I'm losing my hello page, can be disturbing for user, going back one more time without seeing again the hello page and he is out of the app !. But this is another story !)
So for the navigation error, once some binding has been activated by playing with one control, boom!, navigation fails and the menu (the master page) is staying in place. But the most funny is when you click again to navigate, the right page is coming...
This problem can be seen on Android, not on UWP (for iOS I dunno my Mac is shut down at this time...).
In all cases many thanks Brian for this very good Prism adaptation to Xamarin Forms !

**Comment by brianlagunas** [comment]
@odahan You will have to provide a sample app the reproduces your issue.  I cannot help based on your description.  My initial thought is that you have an issue in your code, but there is always a chance there is a bug in Prism for your specific scenario.

**Comment by odahan** [comment]
@brianlagunas No problem I can understand. It was just in case you'll already have an answer (or a clue...). 
I'll try to isolate the problem. thanks.
